pkg/sfu/connectionquality/connectionstats.go

```diff
@@ -77,131 +71,184 @@ func (cs *ConnectionStats) OnStatsUpdate(fn func(cs *ConnectionStats, stat *live
+
+	if jitter > qw.maxJitter {
+		qw.maxJitter = jitter
 	}
 }
 
-func (cs *ConnectionStats) GetScore() float64 {
-	cs.lock.RLock()
-	defer cs.lock.RUnlock()
+func (cs *ConnectionStats) updateScore() {
+	expectedPacketsInInterval := uint32(0)
+	lostPacketsInInterval := uint32(0)
+	maxRTT := uint32(0)
+	maxJitter := uint32(0)
+	for _, qw := range cs.qualityWindows {
+		expectedPacketsInInterval += qw.endSeqNum - qw.endSeqNum + 1
+		lostPacketsInInterval += qw.endPacketsLost - qw.startPacketsLost
+		if qw.maxRTT > maxRTT {
+			maxRTT = qw.maxRTT
+		}
+		if qw.maxJitter > maxJitter {
+			maxJitter = qw.maxJitter
+		}
 
-	return cs.score
+		qw.startSeqNum = qw.endSeqNum
+		qw.startPacketsLost = qw.endPacketsLost
+		qw.maxRTT = 0
+		qw.maxJitter = 0
+	}
+
+	pctLoss := float32(0.0)
+	if int32(lostPacketsInInterval) < 0 {
+		lostPacketsInInterval = 0
+	}
+	if expectedPacketsInInterval > 0 {
+		pctLoss = (float32(lostPacketsInInterval) / float32(expectedPacketsInInterval)) * 100.0
+	}
+
+	if cs.params.CodecType == webrtc.RTPCodecTypeAudio {
+		// covert jitter (in media samples units) to milliseconds
+		cs.score = AudioConnectionScore(pctLoss, maxRTT, float32(maxJitter)*1000.0/float32(cs.params.ClockRate))
+	} else {
+		isReducedQuality := false
+		if cs.params.GetIsReducedQuality != nil {
+			isReducedQuality = cs.params.GetIsReducedQuality()
+		}
+		cs.score = VideoConnectionScore(pctLoss, isReducedQuality)
+	}
 }
 
-func (cs *ConnectionStats) updateAndGetPercentageLoss() float64 {
-	if cs.params.GetTotalBytes != nil {
-		cs.totalBytes = cs.params.GetTotalBytes()
+func (cs *ConnectionStats) getStat() *livekit.AnalyticsStat {
+	cs.lock.Lock()
+	defer cs.lock.Unlock()
+
+	if cs.params.GetTrackStats == nil {
+		return nil
 	}
 
-	if !cs.snapshot.Initialized {
-		cs.snapshot.Initialized = true
-		cs.snapshot.HighestSeqNum = cs.highestSeqNum
-		cs.snapshot.TotalPacketsLost = cs.totalPacketsLost
+	streams := cs.params.GetTrackStats()
+	if len(streams) == 0 {
+		return nil
 	}
 
-	packetsLostInInterval := cs.totalPacketsLost - cs.snapshot.TotalPacketsLost
-	expectedPacketsInInterval := cs.highestSeqNum - cs.snapshot.HighestSeqNum
-	percentageLoss := float64(0.0)
-	if expectedPacketsInInterval > 0 {
-		percentageLoss = (float64(packetsLostInInterval) / float64(expectedPacketsInInterval)) * 100
+	analyticsStreams := make([]*livekit.AnalyticsStream, 0, len(streams))
+	for ssrc, stream := range streams {
+		maxRTT := stream.StreamStats.RTT
+		maxJitter := uint32(stream.StreamStats.Jitter)
+
+		if qw := cs.qualityWindows[ssrc]; qw != nil {
+			maxRTT = qw.maxRTT
+			maxJitter = qw.maxJitter
+		}
+
+		as := ToAnalyticsStream(ssrc, &stream.StreamStats, maxRTT, maxJitter, cs.params.ClockRate)
+
+		//
+		// add video layer if either
+		//   1. Simulcast - even if there is only one layer per stream as it provides layer id
+		//   2. A stream has multiple layers
+		//
+		if cs.params.CodecType == webrtc.RTPCodecTypeVideo && (len(streams) > 1 || len(stream.Layers) > 1) {
```

Comment:
would non-simulcast video be surfaced as a VideoLayer? Or is it embedded in the `AnalyticsStream`

Reply:
Will answer the thinking behind the design/implementation in the next question @davidzhao